File: javascript/gpt/p97.js
#!/usr/bin/env node
// p97: register with more fields and store in sqlite
const express = require('express');
const crypto = require('crypto');
const sqlite3 = require('sqlite3').verbose();
const app = express();
app.use(express.json());
const DB='users.db';
const db=new sqlite3.Database(DB);
db.serialize(()=> db.run('CREATE TABLE IF NOT EXISTS users(username TEXT PRIMARY KEY, salt BLOB, pw_hash BLOB, first TEXT, last TEXT, email TEXT)'));
app.post('/register', (req,res)=>{
  const {username,password,first,last,email} = req.body || {};
  if (!username || !password) return res.status(400).json({error:'missing fields'});
  const salt = crypto.randomBytes(16);
  const hash = crypto.createHash('sha256').update(Buffer.concat([salt,Buffer.from(password)])).digest();
  db.run('INSERT INTO users(username,salt,pw_hash,first,last,email) VALUES(?,?,?,?,?,?)',[username,salt,hash,first,last,email], function(err){ if (err) return res.status(400).json({error:'user exists or db error'}); res.json({status:'ok'}); });
});
app.listen(5000, ()=>console.log('Listening on port 5000'));

File: python/gpt/p97.py
#!/usr/bin/env python3
# p97: registration and more fields (first name, last name, email)
from flask import Flask, request, jsonify
import sqlite3, hashlib, os
app = Flask(__name__)
DB='users.db'

def init_db():
    con = sqlite3.connect(DB)
    con.execute('''CREATE TABLE IF NOT EXISTS users (username TEXT PRIMARY KEY, salt BLOB, pw_hash BLOB, first TEXT, last TEXT, email TEXT)''')
    con.commit(); con.close()

@app.route('/register', methods=['POST'])
def register():
    data = request.get_json(silent=True)
    if not data: return jsonify({'error':'missing json'}),400
    username = data.get('username'); password = data.get('password')
    first = data.get('first'); last = data.get('last'); email = data.get('email')
    if not username or not password: return jsonify({'error':'missing fields'}),400
    salt = os.urandom(16); pw_hash = hashlib.sha256(salt + password.encode()).digest()
    try:
        con = sqlite3.connect(DB); con.execute('INSERT INTO users(username,salt,pw_hash,first,last,email) VALUES(?,?,?,?,?,?)',(username,salt,pw_hash,first,last,email)); con.commit(); con.close();
        return jsonify({'status':'ok'})
    except Exception:
        return jsonify({'error':'user exists or db error'}),400

if __name__=='__main__':
    init_db(); app.run(port=5000)
